File: 0725.分隔链表/0725-分隔链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode less_head(0);
        ListNode more_head(1);
        ListNode* less_ptr = &less_head;
        ListNode* more_ptr = &more_head;
        while(head){
            if (head->val >= x){
                more_ptr->next = head;
                more_ptr = more_ptr->next;
            }
            else{
                less_ptr->next = head;
                less_ptr = less_ptr->next;
            }
            head = head->next;
        }
        less_ptr->next = more_head.next;
        more_ptr->next = NULL;
        return less_head.next;
        
    }
};

File: LeetCode-Python/0725.分隔链表/0725-分隔链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        dummy1 = ListNode(-1)
        dummy2 = ListNode(-1)

        smaller_list = dummy1
        larger_list = dummy2
        p = head
        while p:
            if p.val < x:
                new_node = ListNode(p.val)
                smaller_list.next = new_node
                smaller_list = smaller_list.next
            else:
                new_node = ListNode(p.val)
                larger_list.next = new_node
                larger_list = larger_list.next
            p = p.next
        
        smaller_list.next = dummy2.next
        return dummy1.next
